flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/PlanningConfigurationBuilder.java

```diff
@@ -129,16 +142,48 @@ public Context getContext() {
 	}
 
 	/**
-	 * Creates a configured {@link FrameworkConfig} for a planning session.
-	 *
-	 * @param defaultSchema the default schema to look for first during planning
-	 * @return configured framework config
+	 * Returns the SQL parser config for this environment including a custom Calcite configuration.
 	 */
-	public FrameworkConfig createFrameworkConfig(SchemaPlus defaultSchema) {
+	public SqlParser.Config getSqlParserConfig() {
+		return JavaScalaConversionUtil.toJava(calciteConfig(tableConfig).sqlParserConfig()).orElseGet(() ->
+			// we use Java lex because back ticks are easier than double quotes in programming
+			// and cases are preserved
+			SqlParser
+				.configBuilder()
+				.setLex(Lex.JAVA)
+				.build());
+	}
+
+	private CatalogReader createCatalogReader(
+			boolean lenientCaseSensitivity,
+			String currentCatalog,
+			String currentDatabase) {
+		SqlParser.Config sqlParserConfig = getSqlParserConfig();
+		final boolean caseSensitive;
+		if (lenientCaseSensitivity) {
+			caseSensitive = false;
+		} else {
+			caseSensitive = sqlParserConfig.caseSensitive();
+		}
+
+		SqlParser.Config parserConfig = SqlParser.configBuilder(sqlParserConfig)
+			.setCaseSensitive(caseSensitive)
+			.build();
+
+		return new CatalogReader(
```

Comment:
Maybe this is taken care of somewhere else, but I'm wondering how we tell Calcite to look up in new catalogs that are registered later on in user's session.

Reply:
I would say this is the core difference from the previous PR.

We do not create a separate Calcite schema, but we use the `CatalogManagerCalciteSchema` as the root one (that's why we had to create the `CalciteSchemaBuilder`). At the same time `CatalogManagerCalciteSchema` is just a thin adapter around `CatalogManager`. That means any changes to `CatalogManager` are directly visible by Calcite.

Reply:
Great!